encord/client_metadata_schema.py

```diff
@@ -1,26 +1,83 @@
-from typing import Dict, Optional
+import functools
+from typing import Any, Dict, Optional
 
-import encord.orm.client_metadata_schema as orm
+from encord.exceptions import MetadataValidationError
 from encord.http.v2.api_client import ApiClient
+from encord.orm.client_metadata_schema import (
+    ClientMetadataSchema,
+    ClientMetadataSchemaPayload,
+    ClientMetadataSchemaTypes,
+    validate_embedding,
+)
 
 
-def set_client_metadata_schema_from_dict(api_client: ApiClient, json_dict: Dict[str, orm.ClientMetadataSchemaTypes]):
+@functools.lru_cache(maxsize=128)
+def cached_get_client_metadata_schema(api_client: ApiClient) -> Optional[Dict[str, ClientMetadataSchemaTypes]]:
+    return get_client_metadata_schema(api_client=api_client)
+
+
+def _metadata_validation_error_message_builder(key: str, schema_type: ClientMetadataSchemaTypes, value: Any):
+    message = f"Failed to validate at {key=} against {schema_type=}, where we recieved {value=}"
+    return message
+
+
+def validate_metadata(
+    metadata_schema: Optional[Dict[str, ClientMetadataSchemaTypes]], metadata: Optional[Dict[str, Any]]
+) -> bool:
+    if not metadata_schema or not metadata:
+        return True
+    for key, value in metadata.items():
+        if key not in metadata_schema:
+            pass
+        schema_type = metadata_schema[key]
+        if schema_type == ClientMetadataSchemaTypes.EMBEDDING:
+            if not validate_embedding(value):
+                raise MetadataValidationError(
+                    message=_metadata_validation_error_message_builder(key, schema_type, value)
```

Comment:
A fun python challenge is whether this line can be saved from being duplicated across each element of ClientMetadataSchemaTypes. Depends on key_schema_type, value which we are iterating across.